### PR TITLE
autodiff for rfft/irfft

### PR DESCRIPTION
**Pull request by cofinite**
## Autodiff for `rfft` and `irfft`

Apologies in advance if something is off about the PR, I'm new to this 🥹 

### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.*
- [x] Made sure the book is up to date with changes in this PR.**

*~~Unable to finish due to eventual linker errors (https://github.com/tracel-ai/burn/issues/3591 related?).~~ Has now been successfully run, but seemed to skip the autodiff tests?
**I'm not aware of any changes to be made.

### Related Issues/PRs

Solves https://github.com/tracel-ai/burn/issues/4945

### Changes

Implemented the missing module ops `rfft` and `irfft` in the autodiff backend.

### Testing

`cargo run-checks` was unable to run the tests because the `ndarray` backend has not implemented `rfft` yet (these tests are marked as `ignore` for now). The tests were run manually on `Autodiff<Wgpu>` instead.

### CODE REVIEW

File: crates/burn-backend-tests/tests/autodiff/rfft.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+use super::*;
+use burn_tensor::TensorData;
+use burn_tensor::Tolerance;
+use burn_tensor::signal;
+
+#[test]
+#[ignore = "rfft is not supported for ndarray"]
```
Comment (Sublime12) [inline crates/burn-backend-tests/tests/autodiff/rfft.rs]: to run the tests, you need to remove the ignore

File: crates/burn-backend-tests/tests/autodiff/rfft.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    TensorData::assert_approx_eq::<FloatElem>(
+        &prod.to_data(),
+        &loss.to_data(),
+        Tolerance::default(),
+    );
+}
+
+#[test]
+#[ignore = "rfft is not supported for ndarray"]
```
Comment (Sublime12) [inline crates/burn-backend-tests/tests/autodiff/rfft.rs]: same for here

**Comment by cofinite** [comment]
I've removed the `ignore` on the tests as per your request, though as stated in the OP, they are expected to fail due to the `ndarray` backend not having `rfft` support:
```
---- autodiff::base::rfft::round_trip stdout ----

thread 'autodiff::base::rfft::round_trip' (11032) panicked at crates\burn-ndarray\src\ops\module.rs:387:9:
not yet implemented: rfft is not supported for ndarray

---- autodiff::base::rfft::should_diff_rfft stdout ----

thread 'autodiff::base::rfft::should_diff_rfft' (25908) panicked at crates\burn-ndarray\src\ops\module.rs:387:9:
not yet implemented: rfft is not supported for ndarray
```
It did not seem prudent to attempt using a different backend within the test suite itself, but I manually checked that the tests run successfully on `Autodiff<Wgpu>`.

File: crates/burn-backend-tests/tests/autodiff/rfft.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+use super::*;
+use burn_tensor::TensorData;
+use burn_tensor::Tolerance;
+use burn_tensor::signal;
+
+#[test]
+fn should_diff_rfft() {
```
Comment (laggui) [inline crates/burn-backend-tests/tests/autodiff/rfft.rs:11]: The tests can be feature-gated instead 🙂 iirc the implementation is missing for ndarray and candle, so you can add:

```rust
#[cfg(not(any(feature = "ndarray", feature = "candle")))]
```

File: crates/burn-autodiff/src/ops/module.rs [hunk 4]
```diff
@@ -1973,23 +1977,155 @@ impl<B: Backend, C: CheckpointStrategy> ModuleOps<Autodiff<B, C>> for Autodiff<B
...
+            .prepare::<C>([signal.node.clone()])
+            .compute_bound()
+            .stateful()
+        {
+            OpsKind::Tracked(prep) => prep.finish(state, spectrum),
+            OpsKind::UnTracked(prep) => prep.finish(spectrum),
+        };
+
+        let re = Self::float_slice(spectrum.clone(), &slices_re);
+        let im = Self::float_slice(spectrum.clone(), &slices_im);
```
Comment (laggui) [inline crates/burn-autodiff/src/ops/module.rs:2052]: I think slice / cat should match the FFT `dim` no? Your tests only used dim = 0 so it might only pass in this particular case.

Might be a good idea to add tests for other dimensions too.

File: crates/burn-autodiff/src/ops/module.rs [hunk 5]
```diff
@@ -1973,23 +1977,155 @@ impl<B: Backend, C: CheckpointStrategy> ModuleOps<Autodiff<B, C>> for Autodiff<B
...
+                    grads.register::<B>(node.id, grad_re);
+                }
+                if let Some(node) = node_im {
+                    grads.register::<B>(node.id, grad_im);
+                }
+            }
+        }
+
+        let signal = B::irfft(spectrum_re.primitive, spectrum_im.primitive, dim, n);
+        let n_fft = n.unwrap_or(signal.shape()[dim]);
```
Comment (laggui) [inline crates/burn-autodiff/src/ops/module.rs:2100]: I *think* this is incorrect, and should be inferred from input spectrum size `2 * (spectrum_size - 1)` (when `n` is `None` for `irfft`).

**Review by laggui — CHANGES_REQUESTED** [review]
Thanks for tackling the missing implementation!

File: crates/burn-backend-tests/tests/autodiff/rfft.rs (hunk 3 above)
Comment (cofinite) [reply crates/burn-backend-tests/tests/autodiff/rfft.rs:11]: Done!

File: crates/burn-autodiff/src/ops/module.rs (hunk 4 above)
Comment (cofinite) [reply crates/burn-autodiff/src/ops/module.rs:2052]: It ends up not making a difference which dimension we choose here, because the tensor is immediately after sliced back in half along the same dimension. I've added a test to confirm this.

File: crates/burn-autodiff/src/ops/module.rs (hunk 5 above)
Comment (cofinite) [reply crates/burn-autodiff/src/ops/module.rs:2100]: As I was investigating this, I ran into some issues regarding cases where `n` is `Some` more generally. Will look further into that later.

File: crates/burn-autodiff/src/ops/module.rs (hunk 4 above)
Comment (laggui) [reply crates/burn-autodiff/src/ops/module.rs:2052]: Ahhh sorry you're right, somehow missed this was out of the actual backward implementation, but a workaround for single-output autodiff node (current design limitation) to accept the a split-output operation.

Thanks for adding tests nonetheless!

File: crates/burn-autodiff/src/ops/module.rs (hunk 5 above)
Comment (cofinite) [reply crates/burn-autodiff/src/ops/module.rs:2100]: There's currently a separate issue where if `n` differs from the input signal length, the gradient from `rfft` has the wrong length (`n` instead of the original signal length). We would need to perform the same kind of zero-padding/truncation that `rfft` itself performs, but on the gradient. I'll likely have time to implement this over the weekend.

Comment (cofinite) [reply crates/burn-autodiff/src/ops/module.rs:2100]: The issue is also present in `irfft`. Both ops will have to store the length of their input in the state and then resize the gradient to match.

Comment (cofinite) [reply crates/burn-autodiff/src/ops/module.rs:2100]: > I _think_ this is incorrect, and should be inferred from input spectrum size `2 * (spectrum_size - 1)` (when `n` is `None` for `irfft`).

If `n` is `None`, the output signal length will be the FFT size that was used in the op. The expression you suggested will also result in the same value, so we could use either. I've kept this unchanged for now.

**Comment by cofinite** [comment]
I've added more tests for edge cases, which all run successfully under `Autodiff<Wgpu>`:
```
running 7 tests
test tests::round_trip_with_some_n_greater ... ok
test tests::round_trip_with_dim_nonzero ... ok
test tests::round_trip_inverse_with_some_n_greater ... ok
test tests::round_trip ... ok
test tests::should_diff_rfft ... ok
test tests::round_trip_inverse_with_some_n_less ... ok
test tests::round_trip_with_some_n_less ... ok

test result: ok. 7 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.30s
```
This implementation is, _I think_, future-proof for if/when `rfft`/`irfft` get support for non-power-of-two FFT sizes.

**Review by laggui — APPROVED** [review]
LGTM!

Just had to add a condition to ignore lower float precision in the tests.
